open-craft/openedx-blockstore-relay
# LX-24 Adds management command to copy Units from Open edX to Blockstore

**Description:** Adds a Open edX Django plugin application for transferring data between Open edX and Blockstore.

Initial commit provides a management command which transfers a Unit (vertical) to a Blockstore Bundle.

**JIRA:** LX-24, OC-5133

**Dependencies:**

* https://github.com/edx/edx-platform/pull/18855 -- for static asset parsing
* https://github.com/open-craft/blockstore/pull/4 -- for API
* https://github.com/open-craft/blockstore/pull/5 -- for bundle structure

**Installation instructions:**

1. Install this branch in the CMS shell.
1. Run the Blockstore service so the CMS shell can access its REST API.
1. Ensure that the URL for the Blockstore is present in `settings.BLOCKSTORE_API_URL` (see [`openedx_blockstore_relay.settings`](https://github.com/open-craft/openedx-blockstore-relay/tree/jill/transfer-unit/openedx_blockstore_relay/settings)).

**Testing instructions:**

1. Using the Blockstore REST API, create a Collection, and note its UUID.
   You can pass the `--collection-uuid` to the command below, and a new Bundle will be created for each transfer.
   Alternatively, you can create a Bundle and pass its `--bundle-uuid` instead.
1. Test uploading the various supported types of blocks to the Collection: `course`, `chapter`, `sequential`, `vertical`, `problem`, `video`, `html`.
   Unsupported block types (e.g. `discussion`) are skipped, and show a warning.
   Usage example:
   ```
   ./manage.py cms transfer_to_blockstore --settings=devstack \
      --block-key "block-v1:edX+DemoX+Demo_Course+type@vertical+block@256f17a44983429fb1a60802203ee4e0" \
      --collection-uuid "cccccccc-cccc-cccc-cccc-cccccccccccc"
   ```
1. Visit the Blockstore Bundle API to see the new/updated bundles.
   Each transfer should store in the Bundle:
   * An `.olx` file for the top-level block and each child block
   * All static course assets referenced by the block and its children (with `public=True`).
      HTML blocks also upload a separate `.html` file (with `public=False`).
   * A `bundle.json` with:
      * The `components` list will contain only the top-level `.olx` path.
      * The `type` field will reflect the type of the top-level block.
      * The `assets` and `dependencies` lists will be empty.

**Author Notes & Concerns:**

1. Because we have to upload each block file individually to the bundle, a full snapshot of the Bundle is made at each upload step.
   It would be nice to be able to limit the number of snapshots taken, since these intermediate snapshots will be inherently incomplete and may be invalid.

**Reviewers:**

CC @ormsbee @bradenmacdonald 

**Merge checklist:**

**Post merge:**
      finished.
